### PR TITLE
pipeline: run gstreamer pipeline in separated context

### PR DESCRIPTION
**Pull request by justinjoy**

### CODE REVIEW

File: gaeguli/pipeline.c [hunk 1]
```diff
@@ -164,10 +168,26 @@ gaeguli_pipeline_dispose (GObject * object)
         "GaeguliPipeline reference!");
   }
 
+  if (!self->loop || !g_main_loop_is_running (self->loop)) {
+    g_main_loop_quit (self->loop);
```
Comment (xhaakon) [inline gaeguli/pipeline.c]: If the loop is not running, you quit it? Shouldn't it be the other way around?

File: gaeguli/pipeline.c [hunk 2]
```diff
@@ -164,10 +168,26 @@ gaeguli_pipeline_dispose (GObject * object)
...
+
+  g_clear_pointer (&self->loop, g_main_loop_unref);
+  g_clear_pointer (&self->context, g_main_context_unref);
+
+  if (!self->thread) {
+    if (self->thread != g_thread_self ())
+      g_thread_join (self->thread);
+    else
+      g_thread_unref (self->thread);
+    self->thread = NULL;
```
Comment (xhaakon) [inline gaeguli/pipeline.c]: `thread` is NULL within this scope.

**Review by xhaakon — CHANGES_REQUESTED** [review]

File: gaeguli/pipeline.c (hunk 1 above)
Comment (justinjoy) [reply gaeguli/pipeline.c]: It's a mistake while copy-n-paste.

File: gaeguli/pipeline.c (hunk 2 above)
Comment (justinjoy) [reply gaeguli/pipeline.c]: I tended to set `NULL` to `thread` after `unref()`. Do you mean using `{` and `}` for if-statement?

File: gaeguli/pipeline.c [hunk 3]
```diff
@@ -158,16 +162,34 @@ gaeguli_pipeline_dispose (GObject * object)
         "GaeguliPipeline reference!");
   }
 
+  if (!self->loop && g_main_loop_is_running (self->loop)) {
```
Comment (xhaakon) [inline gaeguli/pipeline.c]: Isn't `if (!self->loop) { stop loop }` still inverted?

I think calling `g_main_loop_is_running ()` is redundant here. It should suffice to do 
```
  if (self->loop) {
    g_main_loop_quit (self->loop);
  }
```

**Review by xhaakon — CHANGES_REQUESTED** [review]

File: gaeguli/pipeline.c (hunk 2 above)
Comment (xhaakon) [reply gaeguli/pipeline.c]: I mean this whole block is inside `if (!self->thread) { do-something-with-my-NULL-thread }`

Comment (justinjoy) [reply gaeguli/pipeline.c]: ah... you pointed inverted mixup. Thank you!

**Review by xhaakon — APPROVED** [review]

**Comment by justinjoy** [comment]
Thank you for good review!

**Pull request merged**